File: Cpp/0779_k_th_symbol_in_grammar.cpp
// https://leetcode.cn/problems/k-th-symbol-in-grammar/

class Solution {
public:
    int kthGrammar(int n, int k) {
        if (n == 1) {
            return 0;
        }

        int tmp = (1 << (n-2));
        if (k <= tmp) {
            return kthGrammar(n-1, k);
        }
        else {
            return 1 - kthGrammar(n-1, k-tmp);
        }
    }
};

File: Python3/0779_k_th_symbol_in_grammar.py
# https://leetcode.cn/problems/k-th-symbol-in-grammar/

class Solution:
    def kthGrammar(self, n, k):
        """
        Inputs:
            n: int
            k: int
        Outputs:
            res: int
        """

        if n == 1:
            return 0
        
        tmp = (1 << (n-2))
        if k <= tmp:
            return self.kthGrammar(n-1, k)
        else:
            return 1 - self.kthGrammar(n-1, k-tmp)